colfetcher: use IndexFetchSpec

RaduBerinde opened the pull request:
#### row: refactor error handling code

This change refactors the `row.Fetcher` error handling code to no
longer rely on the table descriptor. The interesting part is where we
decode key values for a locking error; we now use the `IndexFetchSpec`
and decode the key directly rather than setting up a fake fetcher
through `DecodeRowInfo`.

Release note: None

#### colfetcher: use IndexFetchSpec

This commit converts the cFetcher to use an IndexFetchSpec instead of
table and index descriptors.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75767)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-:-MunIo7CEVWebZJ65FnW:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/colfetcher/cfetcher.go, line 1140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-MunG4uhCshyvY-zzSpc:-MunG4uhCshyvY-zzSpd:b-ww7tpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b0e95501425fdafca88e9bcded6e24f1dd2ea2/pkg/sql/colfetcher/cfetcher.go#L1140)):*
> ```Go
> 
> // processValueSingle processes the given value (of column
> // family.DefaultColumnID), setting values in cf.machine.colvecs accordingly.
> ```

nit: `family.DefaultColumnID` is no more.
___
*[pkg/sql/colfetcher/cfetcher_setup.go, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-MunGOyaANt6bZERrJLP:-MunGOyaANt6bZERrJLQ:b2u1lpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b0e95501425fdafca88e9bcded6e24f1dd2ea2/pkg/sql/colfetcher/cfetcher_setup.go#L36)):*
> ```Go
> 	spec descpb.IndexFetchSpec
> 	// ColIdxMap is a mapping from ColumnID to the ordinal of the corresponding
> 	// column within the cols field.
> ```

nit: something like `s/cols/spec.FetchedColumns/`.
___
*[pkg/sql/colfetcher/cfetcher_setup.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-MunGVyb9nkOYd9QmNCc:-MunGVyb9nkOYd9QmNCd:b-nu3wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b0e95501425fdafca88e9bcded6e24f1dd2ea2/pkg/sql/colfetcher/cfetcher_setup.go#L38)):*
> ```Go
> 	// column within the cols field.
> 	ColIdxMap catalog.TableColMap
> 	// typs are the types of spec.FetchColumns.
> ```

nit: I think `FetchedColumns`.
___
*[pkg/sql/colfetcher/cfetcher_setup.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-MunGxYQCL4d6AGOdtln:-MunGxYQCL4d6AGOdtlo:bsp4tzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b0e95501425fdafca88e9bcded6e24f1dd2ea2/pkg/sql/colfetcher/cfetcher_setup.go#L172)):*
> ```Go
> 	}
> 
> 	// Remapt he post processing spec. away columns that aren't needed.
> ```

nit: `s/Remapt he/Remap the/`, then there is something off in "spec. away".
___
*[pkg/sql/row/errors.go, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-MunEZIo4kOCLAoeiPSj:-MunEZIo4kOCLAoeiPSk:bk48tj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d5bd1725a0df616a37a1660f5ad49b86e610e1b/pkg/sql/row/errors.go#L225)):*
> ```Go
> 	allColumns bool,
> ) (_ catalog.Index, columnNames []string, columnValues []string, _ error) {
> 	// Strip the tenant prefix and pretend to use the system tenant's SQL codec
> ```

nit: the comment needs an update.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
colfetcher: use IndexFetchSpec

This commit converts the cFetcher to use an IndexFetchSpec instead of
table and index descriptors.

Release note: None

RaduBerinde committed:
row: refactor error handling code

This change refactors the `row.Fetcher` error handling code to no
longer rely on the table descriptor. The interesting part is where we
decode key values for a locking error; we now use the `IndexFetchSpec`
and decode the key directly rather than setting up a fake fetcher
through `DecodeRowInfo`.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75767#-:-MuqeE8f4VfHXaXjM0Ng:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
